keps/sig-network/20190125-ingress-api-group.md

```diff
@@ -161,146 +159,157 @@ consistent with the syntax supported by any of the common proxy vendors:
 
 Among cloud providers, there is also inconsistent levels of support
 for regular expression-based path matching. See the load-balancer
-documentation for [aws][aws-re], [gcp][google-re], [azure][azure-re].
+documentation for [AWS][aws-re], [GCP][google-re], [Azure][azure-re].
 
 It is also the case that our [documentation][ingress-docs] (and most
 Ingress providers) treats the path match as a prefix match. For
 example, a narrow interpretation of the specification would require
 all paths to end with `".*$"`.
 
-#### Proposal
-
-1. Clarify behavior of the existing path regex match and identify that use of
-   path regex is non-portable.
-1. Support prefix matching behavior as the portable alternative.
+A detailed discussion of this issue can be found
+[here](https://github.com/kubernetes/ingress-nginx/issues/555).
 
-Change specification of `ingress.spec.rules.http.paths.path` from the current
-text to indicate the non-portability of the value:
+#### Paths proposal
 
-> The matching behavior of Path is implementation specific. Path is either a
-> prefix or regular expression match.
+1. Explicitly state the match mode of the path.
+1. Support the existing implementation-specific behavior.
+1. Support a portable prefix match and future expansion of behavior.
 
-Add a field `ingress.spec.rules.http.paths.pathPrefix`:
+Add a field `ingress.spec.rules.http.paths.pathType` to indicate
+the desired interpretation of the meaning of the `path`:
 
 ```golang
-type HTTPIngressPath struct {
-  ...
-  // PathPrefix matches the HTTP request if the request path begins with
-  // this string. For example, PathPrefix = "/foo" will match "/foo", "/foox",
-  // "/foo/bar".
+ type HTTPIngressPath struct {
+   ...
+  // Path to match against. The interpretation of Path depends on
+  // the value of PathType.
   //
-  // Both Path and PathPrefix cannot be set (non-empty) at the same time.
-  PathPrefix string
-}
-```
-
-Interoperability between v1beta and v1 will work as follows:
+  // Defaults to "/" if empty.
+  //
+  // +Optional
+  Path string
 
-v1beta1 gets the `PathPrefix" field. v1 gets an annotation
-`ingress.kubernetes.io/non-portable-path-semantics`. If we create a v1 object
-from a v1beta1 where `path` was used, we set the annotation. We keep both Path
-and PathPrefix in the API.
+  // PathType determines the interpretation of the Path
+  // matching. PathType can be one of the following values:
+  //
+  // Prefix - matches based on a URL path prefix split
+  //   by '/'. This is not a substring match. For example:
+  //   "/foo/bar" will match "/foo/bar", "/foo/bar/abc", but
+  //   will not match "/foo/barzzz". A Prefix Path of "/" will
+  //   match any paths.
+  //
+  // ImplementationSpecific - interpretation of the Path
+  //   matching is up to the IngressClass.
+  //
+  // Defaults to "Prefix" if not set.
+   //
+  // +Optional
+  PathType string
```

Comment:
Are ingress controllers required to support all path types?  What if an ingress controller doesn’t want to have implementation specific behavior?  Please clarify what the requirements for implementors are here.

Reply:
Added clarification.